src/script/api/script_priorityqueue.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file script_priorityqueue.cpp Implementation of ScriptPriorityQueue. */
+
+#include "../../stdafx.h"
+#include "script_priorityqueue.hpp"
+#include "script_error.hpp"
+#include "../squirrel_helper.hpp"
+#include "../script_instance.hpp"
+#include "../../debug.h"
+
+#include "../../safeguards.h"
+
+
+static bool operator==(const HSQOBJECT &lhs, const HSQOBJECT &rhs)
+{
+	return lhs._type == rhs._type && lhs._unVal.raw == rhs._unVal.raw;
+}
+
+
+ScriptPriorityQueue::~ScriptPriorityQueue()
+{
+	/* Release reference to stored objects. */
+	auto inst = ScriptObject::GetActiveInstance();
+	if (!inst->InShutdown()) {
+		for (auto &i : this->items) inst->ReleaseSQObject(const_cast<HSQOBJECT *>(&i));
+	}
+
+	this->items.clear();
+}
+
+SQInteger ScriptPriorityQueue::Insert(HSQUIRRELVM vm)
+{
+	HSQOBJECT item;
+	int64 priority;
+	sq_resetobject(&item);
+	sq_getstackobj(vm, 2, &item);
+	sq_getinteger(vm, 3, &priority);
+
+	auto inserted = this->items.insert(item);
+	if (inserted.second) {
+		sq_addref(vm, &item); // Keep object alive.
+		this->queue.emplace(priority, item);
+	}
```

Comment:
As written this breaks assumptions that pathfinders are normally making: the expected behavior is that a queue can hold the same item multiple times. The pathfinder separately keeps track of a list of closed items, ie items it doesn't need to revisit. When it pops something from the queue that is already closed it'll ignore that.

It's expected that sometimes an item gets pushed multiple times with different priority values. The lowest priority value is supposed to win in that case, which is handled as per above.

This code blocks the 2nd insert, which means an item cannot be inserted again with a lower priority than the initial one.

Reply:
Well, I've seen both variants for priority queues in practice, and at least the road pathfinder AI lib seems to never inserted an already included item again, but I'll remove the check if you think this better matches the current AI practice.

Reply:
Actually, on second though, unless I'm completely off, A* with an admissible heuristic should never be able to produce a lower cost for an already processed node.
Does any AI/lib implement something different than A*?

Reply:
Right, but a node is processed at the time it is popped from the queue, not on insert.

Reply:
> Right, but a node is processed at the time it is popped from the queue, not on insert.

I don't see how that contradicts @michicc's comment? Higher-cost duplicates of the same node never need to be processed at all, and with an admissible heuristic all duplicates should *be* higher-cost, so there's no reason to insert them.

On the other hand, using an inadmissible heuristic might be useful as a performance tradeoff - given OpenTTD's fairly wide-open maps, the results would probably look okay.